artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java

```diff
@@ -3933,6 +4007,39 @@ void scheduleSlowConsumerReaper(AddressSettings settings) {
       }
    }
 
+   private void enforceRing() {
+      if (ringSize != -1) { // better escaping & inlining when ring isn't being used
+         enforceRing(null, false);
+      }
+   }
+
+   private void enforceRing(MessageReference refToAck, boolean scheduling) {
+      if (getMessageCountForRing() >= ringSize) {
+         refToAck = refToAck == null ? messageReferences.poll() : refToAck;
```

Comment:
If a queue is both LVQ and RING queue, when we poll ref from messageReferences, we should also remove it from LastValueQueue::map.